### PR TITLE
client identities

### PR DESCRIPTION
**Pull request by aaschaer**
Work for allowing client identities to use the CLI by setting GLOBUS_CLI_CLIENT_ID and GLOBUS_CLI_CLIENT_SECRET environment variables.

One thing I haven't fully tested is switching identities without running logout. Once the old client's tokens have expired all should be well, but it might be possible to be in a bad state before then. Not sure how much we care about that, but we could potentially namespace by client id in the db if we wanted to avoid the issue altogether?

### CODE REVIEW

File: changelog.d/20211213_135655_aaschaer_client_login.md
Comment (sirosen) [inline changelog.d/20211213_135655_aaschaer_client_login.md:5]:
```suggestion
* Add support for client credentials for authentication in the Globus CLI
  by setting GLOBUS_CLI_CLIENT_ID and GLOBUS_CLI_CLIENT_SECRET
  environment variables (:pr:`NUMBER`)
** Both variables must be set to enable this behavior
** Tokens generated with client credentials are cached in the current user's home
   directory, but isolated from any user credentials
** With client credentials, `globus login` is invalid, but `globus logout` can be used
   to revoke any cached tokens
```

I think the changelog is a good place to explain usage, since it's where a lot of people will be looking to see the new feature. Above is a suggested first-pass at fleshing out the text a bit more. We can iterate further on this, but I think we want a bit more text to explain.

We'll also want a page on the docs site, probably. But adding it here is still important.

File: src/globus_cli/login_manager/tokenstore.py
Comment (sirosen) [inline src/globus_cli/login_manager/tokenstore.py:126]: What about moving this check to the start of the method?

I see that there's handling for this in the `delete_templated_client` method, and it might be simpler if we assert that this method can _never_ run when `is_client_login() == True`.

File: src/globus_cli/login_manager/tokenstore.py
Comment (sirosen) [inline src/globus_cli/login_manager/tokenstore.py:160]: Right now, this check would discard any `ValueError` and fallthrough if `is_client_login() == False`.
We want to fix that, but I'm also looking at the overall "flow" and wondering if it might be better to leave this method unmodified.

Instead of making `delete_templated_client()` work when `is_client_login() == True`, what about letting it fail, and updating `globus logout` to ensure that we don't end up calling it?
Otherwise, when reading the source for logout, we see a method call which "looks wrong" (because there shouldn't be a client to delete), and it's not immediately obvious that it's a no-op.

File: src/globus_cli/login_manager/manager.py
Comment (sirosen) [inline src/globus_cli/login_manager/manager.py:298]: This change doesn't seem obviously related to the other content of this PR, but I expect that it's needed to get the accurate scope passed to the client creds token callout.

Can you add a comment which explains? I don't think this impacts the non-client-creds case, but it would be good to have internal documentation about how things behave.

**Review by sirosen — COMMENTED** [review]
I want to be extra-careful and test things thoroughly before approving and merging.
However, this all looks to me like it's going to work.

I've left a few comments on things which I recommend changing.

I think it would also be nice to have two additional tests:
- One which runs any real command (e.g. get-identities) with a client login setup
- One test for the retry handler on the "happy path" where the client creds callout works and the subsequent call succeeds

These will likely require new fixture data.

I'm really happy to see the retry handler was possible to write, and doesn't seem to have been too complex or difficult. But I think we need a test for it.

File: src/globus_cli/login_manager/manager.py
Comment (aaschaer) [reply src/globus_cli/login_manager/manager.py:300]: Added a comment explaining

**Review by sirosen — COMMENTED** [review]
I've come back to this and finished my testing. I think this is good to go.

I'd like us to discuss it briefly at our meeting tomorrow prior to merging, however.
If we put this in without adding the two tests I want to add, they might never get added.
That's not catastrophic -- I'd rather release this change with those tests as a TODO item than hold this feature back for a long time -- but if they're within reach, let's get them included.

**Review by sirosen — APPROVED** [review]
Adding a test for the retry handling would be very valuable, but we don't need to block on it. On the strength of manual testing, I'm going to move ahead with this.

**Pull request merged**
